File: app/src/main/java/com/hillal/acc/ui/transactions/EditTransactionFragment.java
package com.hillal.acc.ui.transactions;

import android.app.DatePickerDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.hillal.acc.R;
import com.hillal.acc.data.model.Account;
import com.hillal.acc.data.model.Transaction;
import com.hillal.acc.databinding.FragmentEditTransactionBinding;
import com.hillal.acc.viewmodel.AccountViewModel;
import com.hillal.acc.data.preferences.UserPreferences;
import com.hillal.acc.data.entities.Cashbox;
import com.hillal.acc.viewmodel.CashboxViewModel;
import com.hillal.acc.ui.cashbox.AddCashboxDialog;
import com.hillal.acc.data.repository.CashboxRepository;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.ArrayList;

public class EditTransactionFragment extends Fragment implements com.hillal.acc.ui.cashbox.AddCashboxDialog.OnCashboxAddedListener {
    private FragmentEditTransactionBinding binding;
    private TransactionsViewModel transactionsViewModel;
    private AccountViewModel accountViewModel;
    private UserPreferences userPreferences;
    private long transactionId;
    private long selectedAccountId = -1;
    private final Calendar calendar = Calendar.getInstance();
    private Transaction oldTransaction;
    private final SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy", new Locale("ar"));
    private long selectedCashboxId = -1;
    private long mainCashboxId = -1;
    private List<Cashbox> allCashboxes = new ArrayList<>();
    private CashboxViewModel cashboxViewModel;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        transactionsViewModel = new ViewModelProvider(this).get(TransactionsViewModel.class);
        accountViewModel = new ViewModelProvider(this).get(AccountViewModel.class);
        userPreferences = new UserPreferences(requireContext());
        cashboxViewModel = new ViewModelProvider(this).get(CashboxViewModel.class);
        
        if (getArguments() != null) {
            transactionId = getArguments().getLong("transactionId", -1);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentEditTransactionBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupViews();
        setupListeners();
        loadAccounts();
        loadTransaction();
        setupCashboxDropdown();
    }

    private void setupViews() {
        // Set initial date
        updateDateField();
    }

    private void setupListeners() {
        binding.saveButton.setOnClickListener(v -> updateTransaction());
        binding.cancelButton.setOnClickListener(v -> Navigation.findNavController(requireView()).navigateUp());
        
        // Date picker listener
        binding.dateEditText.setOnClickListener(v -> showDatePicker());
    }

    private void loadAccounts() {
        accountViewModel.getAllAccounts().observe(getViewLifecycleOwner(), accounts -> {
            if (accounts != null && !accounts.isEmpty()) {
                setupAccountDropdown(accounts);
            }
        });
    }

    private void setupAccountDropdown(List<Account> accounts) {
        String[] accountNames = new String[accounts.size()];
        for (int i = 0; i < accounts.size(); i++) {
            accountNames[i] = accounts.get(i).getName();
        }

        ArrayAdapter<String> adapter = new ArrayAdapter<>(
                requireContext(),
                android.R.layout.simple_dropdown_item_1line,
                accountNames
        );

        binding.accountAutoComplete.setAdapter(adapter);
        binding.accountAutoComplete.setOnItemClickListener((parent, view, position, id) -> {
            selectedAccountId = accounts.get(position).getId();
        });
    }

    private void loadTransaction() {
        if (transactionId != -1) {
            transactionsViewModel.getTransactionById(transactionId).observe(getViewLifecycleOwner(), transaction -> {
                if (transaction != null) {
                    oldTransaction = transaction;
                    populateTransactionData(transaction);
                }
            });
        }
    }

    private void populateTransactionData(Transaction transaction) {
        selectedAccountId = transaction.getAccountId();
        
        // Load account name
        accountViewModel.getAccountById(transaction.getAccountId()).observe(getViewLifecycleOwner(), account -> {
            if (account != null) {
                binding.accountAutoComplete.setText(account.getName(), false);
            }
        });

        binding.amountEditText.setText(String.valueOf(transaction.getAmount()));
        binding.descriptionEditText.setText(transaction.getDescription());
        
        // إضافة اسم المستخدم في الملاحظات إذا كانت فارغة
        String notes = transaction.getNotes();
        if (notes == null || notes.isEmpty()) {
            String userName = userPreferences.getUserName();
            if (!userName.isEmpty()) {
                binding.notesEditText.setText(userName);
            }
        } else {
            binding.notesEditText.setText(notes);
        }

        // Set transaction type
        if ("debit".equals(transaction.getType())) {
            binding.radioDebit.setChecked(true);
        } else {
            binding.radioCredit.setChecked(true);
        }

        // Set currency
        if (getString(R.string.currency_yer).equals(transaction.getCurrency())) {
            binding.radioYer.setChecked(true);
        } else if (getString(R.string.currency_sar).equals(transaction.getCurrency())) {
            binding.radioSar.setChecked(true);
        } else {
            binding.radioUsd.setChecked(true);
        }

        // Set date
        calendar.setTimeInMillis(transaction.getTransactionDate());
        updateDateField();
    }

    private void showDatePicker() {
        DatePickerDialog datePickerDialog = new DatePickerDialog(
                requireContext(),
                (view, year, month, dayOfMonth) -> {
                    calendar.set(year, month, dayOfMonth);
                    updateDateField();
                },
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH)
        );
        datePickerDialog.show();
    }

    private void updateDateField() {
        binding.dateEditText.setText(dateFormat.format(calendar.getTime()));
    }

    private void setupCashboxDropdown() {
        cashboxViewModel.getAllCashboxes().observe(getViewLifecycleOwner(), cashboxes -> {
            allCashboxes = cashboxes != null ? cashboxes : new ArrayList<>();
            List<String> names = new ArrayList<>();
            mainCashboxId = -1;
            for (Cashbox c : allCashboxes) {
                names.add(c.name);
                if (mainCashboxId == -1 && (c.name.equals("الرئيسي") || c.name.equalsIgnoreCase("main"))) {
                    mainCashboxId = c.id;
                }
            }
            if (mainCashboxId == -1 && !allCashboxes.isEmpty()) mainCashboxId = allCashboxes.get(0).id;
            names.add("➕ إضافة صندوق جديد...");
            ArrayAdapter<String> adapter = new ArrayAdapter<>(requireContext(), android.R.layout.simple_dropdown_item_1line, names);
            binding.cashboxAutoComplete.setAdapter(adapter);
            // اختيار الصندوق المرتبط بالمعاملة أو الرئيسي
            if (oldTransaction != null && oldTransaction.getCashboxId() != -1) {
                int idx = 0;
                for (int i = 0; i < allCashboxes.size(); i++) {
                    if (allCashboxes.get(i).id == oldTransaction.getCashboxId()) { idx = i; break; }
                }
                binding.cashboxAutoComplete.setText(allCashboxes.get(idx).name, false);
                selectedCashboxId = oldTransaction.getCashboxId();
            } else if (mainCashboxId != -1) {
                int idx = 0;
                for (int i = 0; i < allCashboxes.size(); i++) {
                    if (allCashboxes.get(i).id == mainCashboxId) { idx = i; break; }
                }
                binding.cashboxAutoComplete.setText(allCashboxes.get(idx).name, false);
                selectedCashboxId = mainCashboxId;
            }
            binding.cashboxAutoComplete.setOnItemClickListener((parent, v, position, id) -> {
                if (position == allCashboxes.size()) {
                    // خيار إضافة صندوق جديد
                    openAddCashboxDialog();
                } else {
                    selectedCashboxId = allCashboxes.get(position).id;
                }
            });
        });
    }

    private void openAddCashboxDialog() {
        AddCashboxDialog dialog = new AddCashboxDialog();
        dialog.setListener(this);
        dialog.show(getParentFragmentManager(), "AddCashboxDialog");
    }

    @Override
    public void onCashboxAdded(String name) {
        android.util.Log.d("EditTransactionFragment", "onCashboxAdded called with name: " + name);
        
        // التحقق من الاتصال بالإنترنت
        if (!isNetworkAvailable()) {
            android.util.Log.w("EditTransactionFragment", "No network connection available");
            Toast.makeText(requireContext(), "لا يوجد اتصال بالإنترنت. يرجى التحقق من الاتصال والمحاولة مرة أخرى", Toast.LENGTH_LONG).show();
            return;
        }

        // الحصول على التوكن و baseUrl
        String token = requireContext().getSharedPreferences("auth_prefs", Context.MODE_PRIVATE)
                .getString("token", null);
        
        if (token == null) {
            android.util.Log.w("EditTransactionFragment", "No authentication token found");
            Toast.makeText(requireContext(), "يرجى تسجيل الدخول أولاً", Toast.LENGTH_SHORT).show();
            return;
        }

        android.util.Log.d("EditTransactionFragment", "Token found, proceeding with cashbox addition");

        // إظهار dialog تحميل
        ProgressDialog loadingDialog = new ProgressDialog(requireContext());
        loadingDialog.setMessage("جاري إضافة الصندوق...");
        loadingDialog.setCancelable(false);
        loadingDialog.show();

        // إرسال الصندوق إلى الخادم
        cashboxViewModel.addCashboxToServer("https://malyp.com/", "Bearer " + token, name, 
            new CashboxRepository.CashboxCallback() {
                @Override
                public void onSuccess(Cashbox cashbox) {
                    android.util.Log.d("EditTransactionFragment", "Cashbox added successfully: id=" + cashbox.id + ", name=" + cashbox.name);
                    loadingDialog.dismiss();
                    // سيتم تحديث القائمة تلقائياً عبر LiveData
                    // حدد الصندوق الجديد تلقائياً بعد إضافته
                    binding.cashboxAutoComplete.setText(cashbox.name, false);
                    selectedCashboxId = cashbox.id;
                    Toast.makeText(requireContext(), "تم إضافة الصندوق بنجاح", Toast.LENGTH_SHORT).show();
                }

                @Override
                public void onError(String error) {
                    android.util.Log.e("EditTransactionFragment", "Error adding cashbox: " + error);
                    loadingDialog.dismiss();
                    Toast.makeText(requireContext(), "خطأ في إضافة الصندوق: " + error, Toast.LENGTH_LONG).show();
                }
            });
    }

    private boolean isNetworkAvailable() {
        ConnectivityManager connectivityManager = (ConnectivityManager) requireContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
        return activeNetworkInfo != null && activeNetworkInfo.isConnected();
    }

    private void updateTransaction() {
        if (selectedAccountId == -1) {
            Toast.makeText(requireContext(), "الرجاء اختيار الحساب", Toast.LENGTH_SHORT).show();
            return;
        }

        String amountStr = binding.amountEditText.getText().toString();
        String description = binding.descriptionEditText.getText().toString();
        String notes = binding.notesEditText.getText().toString();
        boolean isDebit = binding.radioDebit.isChecked();
        String currency = getSelectedCurrency();

        if (amountStr.isEmpty()) {
            binding.amountEditText.setError(getString(R.string.error_amount_required));
            return;
        }

        try {
            double amount = Double.parseDouble(amountStr);

            if (oldTransaction == null) {
                Toast.makeText(requireContext(), "حدث خطأ في تحميل بيانات المعاملة الأصلية", Toast.LENGTH_SHORT).show();
                return;
            }

            Transaction transaction = new Transaction();
            transaction.setId(transactionId);
            transaction.setAccountId(selectedAccountId);
            transaction.setAmount(amount);
            transaction.setType(isDebit ? "debit" : "credit");
            transaction.setDescription(description);
            transaction.setNotes(notes);
            transaction.setCurrency(currency);
            transaction.setTransactionDate(calendar.getTimeInMillis());
            transaction.setUpdatedAt(System.currentTimeMillis());
            transaction.setServerId(oldTransaction.getServerId());
            transaction.setWhatsappEnabled(oldTransaction.isWhatsappEnabled());
            transaction.setSyncStatus(0);

            long cashboxIdToSave = (selectedCashboxId != -1) ? selectedCashboxId : mainCashboxId;
            transaction.setCashboxId(cashboxIdToSave);

            transactionsViewModel.updateTransaction(transaction);
            Toast.makeText(requireContext(), R.string.transaction_updated, Toast.LENGTH_SHORT).show();
            Navigation.findNavController(requireView()).navigateUp();
        } catch (NumberFormatException e) {
            binding.amountEditText.setError(getString(R.string.error_invalid_amount));
        }
    }

    private String getSelectedCurrency() {
        int checkedId = binding.currencyRadioGroup.getCheckedRadioButtonId();
        if (checkedId == R.id.radioYer) {
            return getString(R.string.currency_yer);
        } else if (checkedId == R.id.radioSar) {
            return getString(R.string.currency_sar);
        } else {
            return getString(R.string.currency_usd);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
} 

File: app/src/main/java/com/hillal/acc/ui/transactions/EditTransactionFragment.kt
package com.hillal.acc.ui.transactions

import android.app.DatePickerDialog
import android.app.ProgressDialog
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.DatePicker
import android.widget.Toast
import androidx.core.view.OnApplyWindowInsetsListener
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation.findNavController
import com.hillal.acc.R
import com.hillal.acc.data.entities.Cashbox
import com.hillal.acc.data.model.Account
import com.hillal.acc.data.model.Transaction
import com.hillal.acc.data.preferences.UserPreferences
import com.hillal.acc.databinding.FragmentEditTransactionBinding
import com.hillal.acc.ui.cashbox.AddCashboxDialog
import com.hillal.acc.ui.cashbox.AddCashboxDialog.OnCashboxAddedListener
import com.hillal.acc.ui.transactions.CashboxHelper.isNetworkAvailable
import com.hillal.acc.ui.transactions.CashboxHelper.showErrorMessage
import com.hillal.acc.ui.transactions.CashboxHelper.showSuccessMessage
import com.hillal.acc.viewmodel.AccountViewModel
import com.hillal.acc.viewmodel.CashboxViewModel
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

class EditTransactionFragment : Fragment(), OnCashboxAddedListener {
    private var binding: FragmentEditTransactionBinding? = null
    private var transactionsViewModel: TransactionsViewModel? = null
    private var accountViewModel: AccountViewModel? = null
    private var userPreferences: UserPreferences? = null
    private var transactionId: Long = 0
    private var selectedAccountId: Long = -1
    private val calendar: Calendar = Calendar.getInstance()
    private var oldTransaction: Transaction? = null
    private val dateFormat = SimpleDateFormat("dd/MM/yyyy", Locale("ar"))
    private var selectedCashboxId: Long = -1
    private var mainCashboxId: Long = -1
    private var allCashboxes: MutableList<Cashbox> = ArrayList<Cashbox>()
    private var cashboxViewModel: CashboxViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        transactionsViewModel =
            ViewModelProvider(this).get<TransactionsViewModel>(TransactionsViewModel::class.java)
        accountViewModel =
            ViewModelProvider(this).get<AccountViewModel>(AccountViewModel::class.java)
        userPreferences = UserPreferences(requireContext())
        cashboxViewModel =
            ViewModelProvider(this).get<CashboxViewModel>(CashboxViewModel::class.java)

        if (getArguments() != null) {
            transactionId = getArguments()!!.getLong("transactionId", -1)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentEditTransactionBinding.inflate(inflater, container, false)
        return binding!!.getRoot()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            requireActivity().getWindow().setStatusBarColor(Color.TRANSPARENT)
        }
        setupViews()
        setupListeners()
        loadAccounts()
        loadTransaction()
        // setupCashboxDropdown سيتم استدعاؤها من loadTransaction بعد تحميل المعاملة
        // ضبط insets للجذر لرفع المحتوى مع الكيبورد وأزرار النظام
        ViewCompat.setOnApplyWindowInsetsListener(
            view,
            OnApplyWindowInsetsListener { v: View?, insets: WindowInsetsCompat? ->
                var bottom = insets!!.getInsets(WindowInsetsCompat.Type.ime()).bottom
                if (bottom == 0) {
                    bottom = insets.getInsets(WindowInsetsCompat.Type.systemBars()).bottom
                }
                v!!.setPadding(
                    v.getPaddingLeft(),
                    v.getPaddingTop(),
                    v.getPaddingRight(),
                    bottom
                )
                insets
            })
    }

    private fun setupViews() {
        // Set initial date
        updateDateField()
    }

    private fun setupListeners() {
        binding!!.saveButton.setOnClickListener(View.OnClickListener { v: View? -> updateTransaction() })
        binding!!.cancelButton.setOnClickListener(View.OnClickListener { v: View? ->
            findNavController(
                requireView()
            ).navigateUp()
        })


        // Date picker listener
        binding!!.dateEditText.setOnClickListener(View.OnClickListener { v: View? -> showDatePicker() })
    }

    private fun loadAccounts() {
        accountViewModel!!.getAllAccounts()
            .observe(getViewLifecycleOwner(), Observer { accounts: MutableList<Account?>? ->
                if (accounts != null && !accounts.isEmpty()) {
                    setupAccountDropdown(accounts)
                }
            })
    }

    private fun setupAccountDropdown(accounts: MutableList<Account?>) {
        val accountNames = arrayOfNulls<String>(accounts.size)
        for (i in accounts.indices) {
            accountNames[i] = accounts.get(i)!!.getName()
        }

        val adapter = ArrayAdapter<String?>(
            requireContext(),
            android.R.layout.simple_dropdown_item_1line,
            accountNames
        )

        binding!!.accountAutoComplete.setAdapter<ArrayAdapter<String?>?>(adapter)
        binding!!.accountAutoComplete.setOnItemClickListener(AdapterView.OnItemClickListener { parent: AdapterView<*>?, view: View?, position: Int, id: Long ->
            selectedAccountId = accounts.get(position)!!
                .getId()
        })
    }

    private fun loadTransaction() {
        if (transactionId != -1L) {
            transactionsViewModel!!.getTransactionById(transactionId)!!
                .observe(getViewLifecycleOwner(), Observer { transaction: Transaction? ->
                    if (transaction != null) {
                        oldTransaction = transaction
                        populateTransactionData(transaction)
                    }
                })
        }
    }

    private fun populateTransactionData(transaction: Transaction) {
        Log.d("EditTransaction", "Populating transaction data for ID: " + transaction.getId())
        Log.d("EditTransaction", "Transaction cashbox ID: " + transaction.getCashboxId())

        selectedAccountId = transaction.getAccountId()


        // Load account name
        accountViewModel!!.getAccountById(transaction.getAccountId())
            .observe(getViewLifecycleOwner(), Observer { account: Account? ->
                if (account != null) {
                    binding!!.accountAutoComplete.setText(account.getName(), false)
                }
            })

        binding!!.amountEditText.setText(transaction.getAmount().toString())
        binding!!.descriptionEditText.setText(transaction.getDescription())


        // إضافة اسم المستخدم في الملاحظات إذا كانت فارغة
        val notes = transaction.getNotes()
        if (notes == null || notes.isEmpty()) {
            val userName = userPreferences!!.getUserName()
            if (!userName.isEmpty()) {
                binding!!.notesEditText.setText(userName)
            }
        } else {
            binding!!.notesEditText.setText(notes)
        }

        // Set transaction type
        if ("debit" == transaction.getType()) {
            binding!!.radioDebit.setChecked(true)
        } else {
            binding!!.radioCredit.setChecked(true)
        }

        // Set currency
        if (getString(R.string.currency_yer) == transaction.getCurrency()) {
            binding!!.radioYer.setChecked(true)
        } else if (getString(R.string.currency_sar) == transaction.getCurrency()) {
            binding!!.radioSar.setChecked(true)
        } else {
            binding!!.radioUsd.setChecked(true)
        }

        // Set date
        calendar.setTimeInMillis(transaction.getTransactionDate())
        updateDateField()

        setupCashboxDropdown()
    }

    private fun showDatePicker() {
        val datePickerDialog = DatePickerDialog(
            requireContext(),
            DatePickerDialog.OnDateSetListener { view: DatePicker?, year: Int, month: Int, dayOfMonth: Int ->
                calendar.set(year, month, dayOfMonth)
                updateDateField()
            },
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        )
        datePickerDialog.show()
    }

    private fun updateDateField() {
        binding!!.dateEditText.setText(dateFormat.format(calendar.getTime()))
    }

    private fun setupCashboxDropdown() {
        cashboxViewModel!!.getAllCashboxes()
            .observe(getViewLifecycleOwner(), Observer { cashboxes: MutableList<Cashbox?>? ->
                Log.d(
                    "EditTransaction",
                    "Setting up cashbox dropdown with " + (if (cashboxes != null) cashboxes.size else 0) + " cashboxes"
                )
                allCashboxes = if (cashboxes != null) cashboxes.filterNotNull().toMutableList() else ArrayList<Cashbox>()
                val names: MutableList<String?> = ArrayList<String?>()
                mainCashboxId = -1
                for (c in allCashboxes) {
                    names.add(c.name)
                    if (mainCashboxId == -1L && (c.name == "الرئيسي" || c.name.equals(
                            "main",
                            ignoreCase = true
                        ))
                    ) {
                        mainCashboxId = c.id
                        Log.d(
                            "EditTransaction",
                            "Found main cashbox: " + c.name + " (ID: " + c.id + ")"
                        )
                    }
                }
                if (mainCashboxId == -1L && !allCashboxes.isEmpty()) {
                    mainCashboxId = allCashboxes.get(0).id
                    Log.d(
                        "EditTransaction",
                        "Using first cashbox as main: " + allCashboxes.get(0).name + " (ID: " + allCashboxes.get(
                            0
                        ).id + ")"
                    )
                }
                names.add("➕ إضافة صندوق جديد...")
                val adapter = ArrayAdapter<String?>(
                    requireContext(),
                    android.R.layout.simple_dropdown_item_1line,
                    names
                )
                binding!!.cashboxAutoComplete.setAdapter<ArrayAdapter<String?>?>(adapter)


                // اختيار الصندوق المرتبط بالمعاملة أو الرئيسي
                if (oldTransaction != null && oldTransaction!!.getCashboxId() != -1L) {
                    Log.d(
                        "EditTransaction",
                        "Old transaction cashbox ID: " + oldTransaction!!.getCashboxId()
                    )
                    // البحث عن الصندوق المرتبط بالمعاملة
                    var transactionCashbox: Cashbox? = null
                    for (cashbox in allCashboxes) {
                        if (cashbox.id == oldTransaction!!.getCashboxId()) {
                            transactionCashbox = cashbox
                            break
                        }
                    }

                    if (transactionCashbox != null) {
                        binding!!.cashboxAutoComplete.setText(transactionCashbox.name, false)
                        selectedCashboxId = transactionCashbox.id
                        Log.d(
                            "EditTransaction",
                            "Set to old transaction cashbox: " + transactionCashbox.name + " (ID: " + transactionCashbox.id + ")"
                        )
                    } else if (mainCashboxId != -1L && selectedCashboxId == -1L) {
                        // إذا لم يتم العثور على الصندوق المرتبط، استخدم الرئيسي فقط إذا لم يتم اختيار صندوق بعد
                        for (cashbox in allCashboxes) {
                            if (cashbox.id == mainCashboxId) {
                                binding!!.cashboxAutoComplete.setText(cashbox.name, false)
                                selectedCashboxId = mainCashboxId
                                Log.d(
                                    "EditTransaction",
                                    "Set to main cashbox: " + cashbox.name + " (ID: " + cashbox.id + ")"
                                )
                                break
                            }
                        }
                    }
                } else if (mainCashboxId != -1L && selectedCashboxId == -1L) {
                    // إذا لم تكن هناك معاملة قديمة، استخدم الصندوق الرئيسي فقط إذا لم يتم اختيار صندوق بعد
                    for (cashbox in allCashboxes) {
                        if (cashbox.id == mainCashboxId) {
                            binding!!.cashboxAutoComplete.setText(cashbox.name, false)
                            selectedCashboxId = mainCashboxId
                            Log.d(
                                "EditTransaction",
                                "Set to main cashbox (no old transaction): " + cashbox.name + " (ID: " + cashbox.id + ")"
                            )
                            break
                        }
                    }
                }


                // إذا كان المستخدم قد اختار صندوقاً بالفعل، احتفظ باختياره
                if (selectedCashboxId != -1L) {
                    for (cashbox in allCashboxes) {
                        if (cashbox.id == selectedCashboxId) {
                            binding!!.cashboxAutoComplete.setText(cashbox.name, false)
                            Log.d(
                                "EditTransaction",
                                "Keeping user selection: " + cashbox.name + " (ID: " + cashbox.id + ")"
                            )
                            break
                        }
                    }
                }

                binding!!.cashboxAutoComplete.setOnItemClickListener(AdapterView.OnItemClickListener { parent: AdapterView<*>?, v: View?, position: Int, id: Long ->
                    Log.d("EditTransaction", "Cashbox item clicked at position: " + position)
                    if (position == allCashboxes.size) {
                        // خيار إضافة صندوق جديد
                        Log.d("EditTransaction", "Opening add cashbox dialog")
                        openAddCashboxDialog()
                    } else {
                        selectedCashboxId = allCashboxes.get(position).id
                        Log.d(
                            "EditTransaction",
                            "Selected cashbox: " + allCashboxes.get(position).name + " (ID: " + allCashboxes.get(
                                position
                            ).id + ")"
                        )
                        // إزالة رسالة الخطأ إذا كانت موجودة
                        binding!!.cashboxAutoComplete.setError(null)
                    }
                })


                // تفعيل القائمة المنسدلة عند النقر
                binding!!.cashboxAutoComplete.setFocusable(false)
                binding!!.cashboxAutoComplete.setOnClickListener(View.OnClickListener { v: View? -> binding!!.cashboxAutoComplete.showDropDown() })
            })
    }

    private fun openAddCashboxDialog() {
        val dialog = AddCashboxDialog()
        dialog.setListener(this)
        dialog.show(getParentFragmentManager(), "AddCashboxDialog")
    }

    override fun onCashboxAdded(name: String?) {
        Log.d("EditTransactionFragment", "onCashboxAdded called with name: " + name)

        // إظهار dialog تحميل
        val loadingDialog = ProgressDialog(requireContext())
        loadingDialog.setMessage("جاري إضافة الصندوق...")
        loadingDialog.setCancelable(false)
        loadingDialog.show()

        // استخدام CashboxHelper لإضافة الصندوق
        CashboxHelper.addCashboxToServer(
            requireContext(), cashboxViewModel!!, name,
            object : CashboxHelper.CashboxCallback {
                override fun onSuccess(cashbox: Cashbox?) {
                    Log.d(
                        "EditTransactionFragment",
                        "Cashbox added successfully: id=" + (cashbox?.id ?: "null") + ", name=" + (cashbox?.name ?: "null")
                    )
                    loadingDialog.dismiss()
                    // سيتم تحديث القائمة تلقائياً عبر LiveData
                    // حدد الصندوق الجديد تلقائياً بعد إضافته
                    if (cashbox != null) {
                        binding!!.cashboxAutoComplete.setText(cashbox.name, false)
                        selectedCashboxId = cashbox.id
                        Log.d("EditTransaction", "Updated selectedCashboxId to: " + selectedCashboxId)
                        showSuccessMessage(requireContext(), "تم إضافة الصندوق بنجاح")
                    } else {
                        showErrorMessage(requireContext(), "فشل في إضافة الصندوق")
                    }
                }

                override fun onError(error: String?) {
                    Log.e("EditTransactionFragment", "Error adding cashbox: " + error)
                    loadingDialog.dismiss()
                    showErrorMessage(requireContext(), error)
                }
            })
    }

    private val isNetworkAvailable: Boolean
        get() = isNetworkAvailable(requireContext())

    private fun updateTransaction() {
        if (selectedAccountId == -1L) {
            Toast.makeText(requireContext(), "الرجاء اختيار الحساب", Toast.LENGTH_SHORT).show()
            return
        }

        val amountStr = binding!!.amountEditText.getText().toString()
        val description = binding!!.descriptionEditText.getText().toString()
        val notes = binding!!.notesEditText.getText().toString()
        val isDebit = binding!!.radioDebit.isChecked()
        val currency = this.selectedCurrency

        if (amountStr.isEmpty()) {
            binding!!.amountEditText.setError(getString(R.string.error_amount_required))
            return
        }

        // التحقق من اختيار الصندوق
        val selectedCashboxName =
            binding!!.cashboxAutoComplete.getText().toString().trim { it <= ' ' }
        Log.d("EditTransaction", "Selected cashbox name: " + selectedCashboxName)
        Log.d("EditTransaction", "Selected cashbox ID: " + selectedCashboxId)

        if (selectedCashboxName.isEmpty() || selectedCashboxName == "➕ إضافة صندوق جديد...") {
            Toast.makeText(requireContext(), "الرجاء اختيار الصندوق", Toast.LENGTH_SHORT).show()
            binding!!.cashboxAutoComplete.setError("مطلوب اختيار صندوق")
            return
        }

        try {
            val amount = amountStr.toDouble()

            if (oldTransaction == null) {
                Toast.makeText(
                    requireContext(),
                    "حدث خطأ في تحميل بيانات المعاملة الأصلية",
                    Toast.LENGTH_SHORT
                ).show()
                return
            }

            val transaction = Transaction()
            transaction.setId(transactionId)
            transaction.setAccountId(selectedAccountId)
            transaction.setAmount(amount)
            transaction.setType(if (isDebit) "debit" else "credit")
            transaction.setDescription(description)
            transaction.setNotes(notes)
            transaction.setCurrency(currency)
            transaction.setTransactionDate(calendar.getTimeInMillis())
            transaction.setUpdatedAt(System.currentTimeMillis())
            transaction.setServerId(oldTransaction!!.getServerId())
            transaction.setWhatsappEnabled(oldTransaction!!.isWhatsappEnabled())
            transaction.setSyncStatus(0)

            // البحث عن الصندوق المختار من النص المعروض
            var cashboxIdToSave: Long = -1

            Log.d("EditTransaction", "All cashboxes count: " + allCashboxes.size)
            for (cashbox in allCashboxes) {
                Log.d(
                    "EditTransaction",
                    "Checking cashbox: " + cashbox.name + " (ID: " + cashbox.id + ")"
                )
                if (cashbox.name == selectedCashboxName) {
                    cashboxIdToSave = cashbox.id
                    Log.d(
                        "EditTransaction",
                        "Found matching cashbox: " + cashbox.name + " (ID: " + cashbox.id + ")"
                    )
                    break
                }
            }

            Log.d("EditTransaction", "Cashbox ID from name search: " + cashboxIdToSave)


            // إذا لم يتم العثور على الصندوق من النص، استخدم selectedCashboxId
            if (cashboxIdToSave == -1L && selectedCashboxId != -1L) {
                cashboxIdToSave = selectedCashboxId
                Log.d("EditTransaction", "Using selectedCashboxId: " + selectedCashboxId)
            }


            // إذا لم يتم العثور على صندوق محدد، استخدم الصندوق الرئيسي
            if (cashboxIdToSave == -1L && mainCashboxId != -1L) {
                cashboxIdToSave = mainCashboxId
                Log.d("EditTransaction", "Using mainCashboxId: " + mainCashboxId)
            }


            // إذا لم يكن هناك صندوق رئيسي، استخدم أول صندوق متاح
            if (cashboxIdToSave == -1L && !allCashboxes.isEmpty()) {
                cashboxIdToSave = allCashboxes.get(0).id
                Log.d("EditTransaction", "Using first available cashbox: " + allCashboxes.get(0).id)
            }


            // إذا لم تكن هناك صناديق على الإطلاق، احتفظ بالصندوق القديم
            if (cashboxIdToSave == -1L) {
                cashboxIdToSave = oldTransaction!!.getCashboxId()
                Log.d(
                    "EditTransaction",
                    "Using old transaction cashbox: " + oldTransaction!!.getCashboxId()
                )
                if (cashboxIdToSave == -1L) {
                    Toast.makeText(
                        requireContext(),
                        "خطأ: لا توجد صناديق متاحة",
                        Toast.LENGTH_SHORT
                    ).show()
                    return
                }
            }


            // التحقق النهائي من أن الصندوق صحيح
            if (cashboxIdToSave == -1L) {
                Toast.makeText(requireContext(), "خطأ: لم يتم تحديد صندوق صحيح", Toast.LENGTH_SHORT)
                    .show()
                return
            }

            Log.d("EditTransaction", "Final cashbox ID to save: " + cashboxIdToSave)
            transaction.setCashboxId(cashboxIdToSave)

            transactionsViewModel!!.updateTransaction(transaction)
            Toast.makeText(requireContext(), R.string.transaction_updated, Toast.LENGTH_SHORT)
                .show()
            findNavController(requireView()).navigateUp()
        } catch (e: NumberFormatException) {
            binding!!.amountEditText.setError(getString(R.string.error_invalid_amount))
        }
    }

    private val selectedCurrency: String
        get() {
            val checkedId = binding!!.currencyRadioGroup.getCheckedRadioButtonId()
            if (checkedId == R.id.radioYer) {
                return getString(R.string.currency_yer)
            } else if (checkedId == R.id.radioSar) {
                return getString(R.string.currency_sar)
            } else {
                return getString(R.string.currency_usd)
            }
        }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }
}